src/common/column_matrix.h

```diff
@@ -103,96 +110,179 @@ class ColumnMatrix {
     size_t accum_row_ind_ = 0;
     for (int32_t fid = 0; fid < nfeature; ++fid) {
       boundary_[fid].index_begin = accum_index_;
-      boundary_[fid].row_ind_begin = accum_row_ind_;
       if (type_[fid] == kDenseColumn) {
         accum_index_ += static_cast<size_t>(nrow);
-        accum_row_ind_ += static_cast<size_t>(nrow);
       } else {
         accum_index_ += feature_counts_[fid];
         accum_row_ind_ += feature_counts_[fid];
       }
       boundary_[fid].index_end = accum_index_;
-      boundary_[fid].row_ind_end = accum_row_ind_;
     }
 
-    index_.resize(boundary_[nfeature - 1].index_end);
-    row_ind_.resize(boundary_[nfeature - 1].row_ind_end);
+    if ( (gmat.max_num_bins_ - 1) <= static_cast<int>(std::numeric_limits<uint8_t>::max()) ) {
+      type_size_ = 1;
+    } else if ((gmat.max_num_bins_ - 1) <= static_cast<int>(std::numeric_limits<uint16_t>::max())) {
+      type_size_ = 2;
+    } else {
+      type_size_ = 4;
+    }
 
-    // store least bin id for each feature
-    index_base_.resize(nfeature);
-    for (int32_t fid = 0; fid < nfeature; ++fid) {
-      index_base_[fid] = gmat.cut.Ptrs()[fid];
+    index_.resize(boundary_[nfeature - 1].index_end * type_size_, 0);
+    if (!all_dense) {
+      row_ind_.resize(boundary_[nfeature - 1].index_end);
     }
 
+    // store least bin id for each feature
+    index_base_ = const_cast<uint32_t*>(gmat.cut.Ptrs().data());
+
     // pre-fill index_ for dense columns
+    const bool noMissingValues = gmat.row_ptr[nrow] == nrow * nfeature;
 
-    #pragma omp parallel for
-    for (int32_t fid = 0; fid < nfeature; ++fid) {
-      if (type_[fid] == kDenseColumn) {
-        const size_t ibegin = boundary_[fid].index_begin;
-        uint32_t* begin = &index_[ibegin];
-        uint32_t* end = begin + nrow;
-        std::fill(begin, end, std::numeric_limits<uint32_t>::max());
-        // max() indicates missing values
+    if (noMissingValues) {
+      missing_flags_.resize(boundary_[nfeature - 1].index_end, false);
+    } else {
+      missing_flags_.resize(boundary_[nfeature - 1].index_end, true);
+    }
+
+    if (all_dense) {
+      switch (gmat.index.getBinBound()) {
```

Comment:
Switching by `type_size_` allows to use memory reduction for sparse DMatrix also (`gmat.IsDense() == false`). For sparse DMatrix `gmat.index.getBinBound()` returns `UINT32_BINS_TYPE` always.
As I pointed above: for `GHistIndexMatrix` there is no reason to store bins in `index` in form `[0,max_num_bins - 1]` as we should keep feature index for each bin to select right `offset` (in old version it's `disp`).
And by this reason in case of sparse DMatrix  we should use `uint32_t` always and `offsets`(in previous commit it's `disp`) is `nullptr` for it case.
But still we have a chance to reduce memory for `ColumnMatrix` and that`s why I separate it.

Comment was added in code.